---
title: "GTex_DESeq2_DEAnalysis"
author: "Nickie Safarian"
date: "11/29/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This R Markdown document contains codes for DE analysis of GTex_phs000424 dataset using DESeq2 package.

```{r, include=FALSE}

# devtools::install_github('oganm/markerGeneProfile')

library(tidyverse)
library(magrittr)
library(cowplot)
library(markerGeneProfile)
library(readxl)
library(ggpubr)
library(knitr)
library(DESeq2)

```

## Imort the Metadata

The following chunks read in an older version (V8) of expression counts and metadata:

```{r, echo=FALSE, include=FALSE}

# 1) get sample.metadata from gtex website
gtex_samples = read_delim(url("https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt"), delim = '\t')


# 2) get phenotype.metadata (this file contains Age and sex info for samples)
gtex_pheno_meta = read_delim(url('https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt'), delim = '\t')


```

## Preprocess the Metadata

```{r}

# get samples from specific brain regions (i.e., cortex)
use_cortex_regions = c("Brain - Frontal Cortex (BA9)")

# filter sample.metadata to only include the region of interest
use_samples = gtex_samples %>% filter(SMTSD %in% use_cortex_regions) %>% pull(SAMPID) #425 samples are from BA9

# subset the metadata for only BA9 samples
brain_gtex_metadata = gtex_samples %>% filter(SAMPID %in% use_samples) 

```

In order to combine the samples.metadata and phenotype.metadata, we need to match the ID columns between the two data sets. To do so

```{r}

#  add a column called SUBJID to metadata that enables linking with subject metadata
gtex_samples = gtex_samples %>% 
  separate(SAMPID, into = c('first', 'second'), remove = F, sep = '-') %>% 
  unite(col = "SUBJID", first:second, sep = '-', remove = F) 


# add age and sex column to the previous metadata
METADATA <- merge(gtex_pheno_meta, brain_gtex_metadata, 
                  by.x='SUBJID', by.y = 'SUBJID') #425 IDs

rownames(METADATA) <- METADATA$SAMPID

```

## Import the counts matrix

```{r, include=FALSE}

# read in full RNAseq reads - careful, this file is  huge (14 GB)
reads_data_file_name = '/external/rprshnas01/netdata_kcni/stlab/Public/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct'

gtex_reads = read_tsv(paste0(reads_data_file_name), skip=2)


kable(gtex_reads[1:3, 1:4]) # 56200 * 17384 

```


```{r}

# subset counts/reads for the select samples
get_sample_col_names = intersect(c('Name', 'Description', use_samples), colnames(gtex_reads))

brain_gtex_reads = gtex_reads[,get_sample_col_names] # 211 cols

# remove Ensembl IDs from first column and set as rownames
brain_gtex_reads = brain_gtex_reads %>% column_to_rownames(var= 'Name')
brain_gtex_reads = brain_gtex_reads[, -1] # remove description coulmn, 209 cols

```


#------------------------------------------------------------------------------

## DESEq2 steps

```{r, include=FALSE}

# match Ids between metadata and counts
coldata = METADATA%>% subset(SAMPID %in% colnames(brain_gtex_reads)) #209

# Very important:
cts <- round(brain_gtex_reads) %>%  # convert values to integers
  as.data.frame() %>%
  dplyr::filter(rowSums(.) > 0) # remove genes with zero total 
                                 # counts across samples

```

### Check coldata & Cts IDs are matching and in the same order

```{r}

# Ensure coldata and cts are in the same order
all(rownames(coldata) == colnames(cts))
# Ensure the same samples across cts and coldata
all(rownames(coldata) %in% colnames(cts))

```

### Prepare covariates in the coldata accordingly

```{r}

# we have to remove "-" character which sits between the age ranges, as DESeq2 gets confused by it.
coldata$AGE <- gsub("-", "_", coldata$AGE)


coldata$AGE = factor(coldata$AGE, c('20_29', '30_39', '40_49', '50_59', '60_69', '70_79')) # age is set in range in this dataset

coldata$SEX <- factor(coldata$SEX , c("1","2"))

```


```{r}

dds <- DESeqDataSetFromMatrix(countData = cts, 
                              colData = coldata, 
                              design = ~ SEX+AGE)
```

```{r}

# Perform pre-filtering to remove genes that are lowly expressed 
# (at least 10 counts)
dds <- estimateSizeFactors(dds)
ids <- rowSums(counts(dds, normalized=TRUE) >= 10 ) >= 3 
dds <- dds[ids, ]

```

```{r}

dds$AGE <- relevel(dds$AGE, '20_29')
dds$SEX <- relevel(dds$SEX, "1")

```

```{r}

dds <- DESeq(dds)
resultsNames(dds) 
GTex.Res.Age <- results(dds, name="AGE_70_79_vs_20_29", alpha = 0.05)
summary(GTex.Res.Age)
 
```







```{r}

### perform normalization for gene expression data from GTEX

gtex_brain_expr_cpm = edgeR::cpm(brain_gtex_reads[, intersect(use_samples, colnames(brain_gtex_reads))], prior.count = 1)

rownames(gtex_brain_expr_cpm) = str_extract(brain_gtex_reads$Name, 'ENSG\\d+')

gtex_brain_expr_cpm = gtex_brain_expr_cpm %>% as.data.frame() %>% tibble::rownames_to_column(var = "ensembl_id")

```



```{r}

# marker genes list (importing and processing)

markers_df <- read.csv(url('https://raw.githubusercontent.com/sonnyc247/MarkerSelection/master/Data/Outputs/CSVs_and_Tables/Markers/All_hodge_regions/new_ALLReg_results_SubclassWithExc.csv'))

cell_types = markers_df$subclass %>% unique()
marker_list = lapply(cell_types, function(cell_type){
  return(markers_df %>% filter(subclass == cell_type) %>% pull(gene) %>% unlist())
})
names(marker_list) = cell_types

```














